### PR TITLE
change where transformers are called

### PR DESCRIPTION
**Pull request by lavalamp**
#### What type of PR is this?

/kind bug

/kind regression

#### What this PR does / why we need it:

Code changes needed for the problem discovered in https://github.com/kubernetes/kubernetes/pull/116432

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/client-go/tools/cache/delta_fifo.go
Comment (smarterclayton) [inline staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:146]: duplicate godoc

**Comment by smarterclayton** [comment]
Looks correct to me at first pass, but I will defer to someone who has looked at deltafifo more recently.

**Comment by liggitt** [comment]
I don't think I've ever looked at the transforming informer code before... I nominate @wojtek-t @deads2k as reviewers based on https://github.com/kubernetes/kubernetes/pull/104300

File: staging/src/k8s.io/client-go/tools/cache/delta_fifo.go
Comment (lavalamp) [reply staging/src/k8s.io/client-go/tools/cache/delta_fifo.go]: heh, so was the original.

I updated all the comments to reflect my current beliefs about how it behaves now. It needs tests to verify.

File: staging/src/k8s.io/client-go/tools/cache/delta_fifo.go
Comment (odinuge) [inline staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:147]: Nice, this comment is very good!

Should we maybe rephrase this to state that "ignore" means returning the input object to be 100% clear?

Comment (lavalamp) [reply staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:147]: Yes, I will update. I'm adding a test now

**Comment by odinuge** [comment]
I think this makes a lot of sense since we keep the old behavior of sending `DeletedFinalStateUnknown` through the conversion func as well, so that clients depending on unpacking it inside of this func will still work as expected. And as you say, the `DeletedFinalStateUnknown` will always have an Obj that is already converted.


I was thinking about maybe doing it inside the reflector or a ListWatcher override before sending it to the deltafifo , but that would "break" those who depend on getting the `DeletedFinalStateUnknown` into the conversion func and unpacking it.

**Comment by odinuge** [comment]
What do you think about backporting this PR together with https://github.com/kubernetes/kubernetes/pull/115620?

File: staging/src/k8s.io/client-go/tools/cache/delta_fifo.go
Comment (odinuge) [reply staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:147]: Nice, thanks! Also feel free to take some logic from https://github.com/kubernetes/kubernetes/pull/116432 as well, or we can do that as a follow up after this, thats up to you

**Comment by lavalamp** [comment]
I ran unit tests for an hour with stress and race and no problems found.

> What do you think about backporting this PR together with https://github.com/kubernetes/kubernetes/pull/115620?

Yeah they need to be backported together. But right now I'm worried about not blocking the release of 1.27 :)

/milestone v1.27

**Comment by lavalamp** [comment]
@odinuge do you think you can rebase https://github.com/kubernetes/kubernetes/pull/116432 on top of this so we can use the test you wrote there? I think after this PR there's a case that you shouldn't have to handle any more (tombstone object with untransformed inner object).

**Review by odinuge — APPROVED** [review]
> @odinuge do you think you can rebase https://github.com/kubernetes/kubernetes/pull/116432 on top of this so we can use the test you wrote there? I think after this PR there's a case that you shouldn't have to handle any more (tombstone object with untransformed inner object).

Sweet. Did that now, and the tests passed the race detector locally without any problems. I'll clean that PR up when this is merged and I have some more time maybe tomorrow.


Overall I like this approach, and the code looks good, together with a very descriptive set of godocs. Thanks!

/lgtm

Adding a hold in case others want to look, but anyone feel free to remove it!
/hold

**Comment by wojtek-t** [comment]
For posterity - this LGTM too

Thanks for fixing this Daniel!

**Pull request merged**
